typing/typedecl.ml

```diff
@@ -226,6 +253,34 @@ let make_constructor env type_path type_params sargs sret_type =
       widen z;
       targs, Some tret_type, args, Some ret_type
 
+(* Check that the argument to a GADT constructor is compatible with unboxing
+   the type, given the existential variables introduced by this constructor. *)
+let rec check_unboxed_gadt_arg loc ex env ty =
+  match get_unboxed_type_representation env ty with
+  | Some {desc = Tvar _ as d; _} ->
+    let f t = (Btype.repr t).desc = d in
+    if List.exists f ex then
+      raise(Error(loc, Wrong_unboxed_type
+                    "it might contain both float and non-float values"))
+  | Some {desc = Tarrow _ | Ttuple _ | Tpackage _ | Tobject _ | Tnil
+                 | Tvariant _; _} ->
+    ()
+    (* A comment in [Translcore.transl_exp0] claims the above cannot be
+       represented by floats. *)
+  | Some {desc = Tconstr (p, args, _); _} ->
+    let tydecl = Env.find_type p env in
+    assert (not (Builtin_attributes.has_unboxed tydecl.type_attributes));
+    if tydecl.type_kind = Type_abstract then
+      List.iter (check_unboxed_gadt_arg loc ex env) args
+  | Some {desc = Tfield _ | Tlink _ | Tsubst _; _} -> assert false
+      (* TODO: check with Jacques that Tfield is indeed impossible *)
+  | Some {desc = Tunivar _; _} -> ()
```

Comment:
If I understand `Tunivar` correctly, you can only get it if you do some variant of this:
```
type t = A of {'a . 'a} [@@unboxed]
```
In this case it is safe because the argument must have all types at the same time, which means it cannot be a float (or anything else).
But I would like to get @garrigue's answer to this question: can a universal type variable's scope range over a type definition?